File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/modloaders/OptiFineUtils.java
package net.kdt.pojavlaunch.modloaders;

import net.kdt.pojavlaunch.utils.DownloadUtils;

import java.util.List;

public class OptiFineUtils {

    /**
     * OptiFineの利用可能なバージョン一覧をダウンロードします。
     * @param force キャッシュを無視する場合はtrue
     * @return OptiFineバージョン情報
     * @throws Exception ダウンロードまたはパースに失敗した場合
     */
    public static OptiFineVersions downloadOptiFineVersions(boolean force) throws Exception {
        return DownloadUtils.downloadStringCached("https://optifine.net/downloads", "of_downloads_page", force, new OptiFineScraper());
    }

    /**
     * OptiFineのバージョン情報を保持するクラス
     */
    public static class OptiFineVersions {
        public List<String> minecraftVersions;
        public List<List<OptiFineVersion>> optifineVersions;
    }

    /**
     * 個別のOptiFineバージョンを表すクラス
     */
    public static class OptiFineVersion {
        public String minecraftVersion;
        public String versionName;
        public String downloadUrl;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/feature/mod/modloader/OptiFineUtils.kt
package com.arata.yukarilauncher.feature.mod.modloader

import com.arata.yukarilauncher.utils.http.DownloadUtils

/** OptiFineのバージョン情報を管理するユーティリティオブジェクト。 */
object OptiFineUtils {

    /** OptiFineのバージョン一覧をダウンロードして取得する。 @param force キャッシュを無視して再取得する場合はtrue @return OptiFineのバージョン情報 */
    fun downloadOptiFineVersions(force: Boolean): OptiFineVersions {
        return DownloadUtils.downloadStringCached("https://optifine.net/downloads", "of_downloads_page", force, OptiFineScraper())
    }

    /** OptiFineの全バージョン情報を保持するデータクラス。 */
    class OptiFineVersions {
        /** Minecraftバージョンのリスト。 */
        var minecraftVersions: MutableList<String>? = null
        /** MinecraftバージョンごとのOptiFineバージョンリスト。 */
        var optifineVersions: MutableList<List<OptiFineVersion>>? = null
    }

    /** 個別のOptiFineバージョン情報を保持するデータクラス。 */
    class OptiFineVersion {
        /** 対応するMinecraftバージョン。 */
        var minecraftVersion: String? = null
        /** OptiFineのバージョン名。 */
        var versionName: String? = null
        /** ダウンロードURL。 */
        var downloadUrl: String? = null
    }
}
